native/cocos/renderer/pipeline/shadow/CSMLayers.cpp

```diff
@@ -0,0 +1,300 @@
+        const float modY = fmodf(projPos.y, texelSize.y);
+        const Vec3 projSnap(projPos.x - modX, projPos.y - modY, projPos.z);
+        const Mat4 matShadowViewProjArbitaryPosInv = matShadowViewProjArbitaryPos.getInversed();
+        Vec3 snap;
+        snap.transformMat4(projSnap, matShadowViewProjArbitaryPosInv);
+
+        Mat4::fromRT(dirLight->getNode()->getRotation(), snap, &matShadowTrans);
+        matShadowView = matShadowTrans.getInversed();
+
+        // fill data
+        Mat4 matShadowViewProj;
+        Mat4::multiply(matShadowProj, matShadowView, &matShadowViewProj);
+        _matShadowView = matShadowView.clone();
+        _matShadowProj = matShadowProj.clone();
+        _matShadowViewProj = matShadowViewProj.clone();
+    }
+    _validFrustum.createOrtho(orthoSizeWidth, orthoSizeHeight, 0.1F, _shadowCameraFar, matShadowTrans);
+}
+
+CSMLayerInfo::CSMLayerInfo(uint level) : ShadowTransformInfo(level) {
+    _splitCameraNear = 0.0F;
+    _splitCameraFar = 0.0F;
+    _matShadowAtlas.setZero();
+    _matShadowViewProjAtlas.setZero();
+    calculateAtlas(level);
+}
+
+void CSMLayerInfo::calculateAtlas(uint level) {
+    const gfx::Device *device = gfx::Device::getInstance();
+    const float clipSpaceSignY = device->getCapabilities().clipSpaceSignY;
+    const float x = floorf(static_cast<float>(level % 2U)) - 0.5F;
+    const float y = clipSpaceSignY * (0.5F - floorf(static_cast<float>(level) / 2U));
+    const Vec3 bias(x, y, 0.0F);
+    const Vec3 scale(0.5F, 0.5F, 1.0F);
+    Mat4::fromRTS(Quaternion::identity(), bias, scale, &_matShadowAtlas);
+}
+
+CSMLayers::CSMLayers() {
+    _levelCount = 0U;
+    _specialLayer = new ShadowTransformInfo(1U);
+    _shadowDistance = 0.0F;
+}
+
+CSMLayers::~CSMLayers() {
+    if (_specialLayer) {
+        delete _specialLayer;
+        _specialLayer = nullptr;
+    }
+
+    for (const auto *shadowCSMLayer : _layers) {
+        if (shadowCSMLayer) {
+            delete shadowCSMLayer;
+            shadowCSMLayer = nullptr;
+        }
+    }
+
+    _layers.clear();
+}
+
+void CSMLayers::update(const PipelineSceneData *sceneData, const scene::Camera *camera) {
+    CC_ASSERT(sceneData);
+    CC_ASSERT(camera);
+
+    const scene::Shadows *shadowInfo = sceneData->getShadows();
+    const scene::RenderScene *const scene = camera->getScene();
+    scene::DirectionalLight *dirLight = scene->getMainLight();
+
+    CC_ASSERT(dirLight);
+
+    const uint levelCount = static_cast<uint>(dirLight->getShadowCSMLevel());
+    const float shadowDistance = dirLight->getShadowDistance();
+
+    if (!shadowInfo->isEnabled() || !dirLight->isShadowEnabled()) {
+        return;
+    }
+
+    if (dirLight->isShadowFixedArea()) {
+        updateFixedArea(dirLight);
+    } else {
+        bool isRecalculate = false;
+        for (uint i = 0; i < levelCount; ++i) {
+            if (_layers.size() <= i) {
+                _layers.emplace_back(new CSMLayerInfo(i));
```

Comment:
Please use `ccnew` instead.

Reply:
And it is better to `std::vector::reserve()` it outside for. The code can change to
```c++
size_t oldLayerSize = _layers.size;
bool isRecalculate = oldLayerSize <= levelCount;
_layers.reserve(levelCount);
while (oldLayerSize++ < levelCount) {
    _layers.emplace_back(ccnew CSMLayer(oldLayerSize));
}
```

Reply:
The loop logic is wrong. I can't modify it like this at present

Reply:
Why, i think i didn't change the logic.

Reply:
* levelCount  range:[1~4]
* _layers[]  index;[0~3]

Index mismatch

Reply:
Then can reserve levelCount + 1.

Reply:
if oldLayerSize <= levelCount  is all the true, In other words, the next question of `splitFrustumLevels()` has to be recalculated all the time